File: src/main/java/com/ultreon/ultranlang/NodeVisitor.java
package com.ultreon.ultranlang;

import com.ultreon.ultranlang.annotations.Visit;
import com.ultreon.ultranlang.ast.AST;
import org.jetbrains.annotations.Nullable;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;

public class NodeVisitor {
    @Nullable
    public Object visit(AST node) {
        for (Method method : this.getClass().getMethods()) {
            for (Annotation annotation : method.getAnnotations()) {
                if (annotation.annotationType().equals(Visit.class)) {
                    if (method.getParameterTypes()[0].isAssignableFrom(node.getClass())) {
                        try {
                            return method.invoke(this, node);
                        } catch (Exception e) {
                            e.printStackTrace();
                            System.exit(1);
                        }
                    } else {
                    }
                } else {
                }
            }
        }

        throw new IllegalArgumentException("No visitor found for " + node.getClass().getName());
    }
}

File: src/main/kotlin/com/ultreon/ultranlang/NodeVisitor.kt
package com.ultreon.ultranlang

import com.ultreon.ultranlang.annotations.Visit
import com.ultreon.ultranlang.ast.AST
import java.lang.IllegalStateException

open class NodeVisitor {
    fun visit(node: AST): Any? {
        for (member in this::class.members) {
            for (annotation in member.annotations) {
                if (annotation is Visit) {
                    if (annotation.value.isInstance(node)) {
                        return member.call(this, node)
                    }
                }
            }
        }
        throw IllegalStateException("No visitor found for ${node::class.qualifiedName}")
    }
}